cpp/src/arrow/engine/substrait/expression_internal.h

```diff
@@ -34,6 +35,15 @@ namespace engine {
 
 class SubstraitCall;
 
+ARROW_ENGINE_EXPORT
+Result<FieldRef> DirectReferenceFromProto(const substrait::Expression::ReferenceSegment*,
+                                          const ExtensionSet&, const ConversionOptions&);
+
+ARROW_ENGINE_EXPORT
+Result<compute::Expression> FromProto(const substrait::Expression::ReferenceSegment*,
+                                      const ExtensionSet&, const ConversionOptions&,
+                                      std::optional<compute::Expression>);
```

Comment:
It's the "current expression".  This method is currently called int he middle of a deserializing an expression tree.  So, for example:

```mermaid
flowchart TD
    A[Call] -->|args| FieldRef
    A -->|args| C
    C[Call*] -->|args| D
    C -->|Call| E
    D[Literal]
    E[FieldRef*]
```

So, when de-referencing `FieldRef` this will be `Call` and when dereferencing `FieldRef*` this will be `Call*`.

However, can we remove this prototype from the header file and put it in an anonymous namespace?  I think it should be an internal method and not exposed.

Reply:
Thanks Weston for the explanation.

+1 on not exposing this in the header file. I think if we move the MakeSegmentedAggregateRel to expression_internal.cc then we don't need to expose this via the header file. I have a comment about moving these "make extension rel" methods out of options.cc anyway.

Reply:
I don't think `expression_internal.cc`, which deals with Substrait expression parsing, is the right place for any of the `Make*Rel` methods, which deal with Substrait extension handling.

IMO, refactoring of `options.cc` could be reasonable but is outside the scope of this PR.

Reply:
> However, can we remove this prototype from the header file and put it in an anonymous namespace? I think it should be an internal method and not exposed.

@westonpace Would sth like this look better  to you? https://github.com/apache/arrow/commit/e958b6d6cfd8889fd8e5a616eadfe0023cac3b77

Reply:
> @westonpace Would sth like this look better to you? https://github.com/apache/arrow/commit/e958b6d6cfd8889fd8e5a616eadfe0023cac3b77

I'm neutral on the change and don't want to spend too much time bike shedding here.  If you're comfortable let's move forward.

> DirectReferenceFromProto is used by MakeSegmentedAggregateRel in options.cc, so I think it has to be in a header.

Yes, I am fine with `DirectReferenceFromProto`.  I was merely asking about the `FromProto` variant that takes a reference segment.